### DISCUSSION TITLE
Make common util directive logic generic?

### DISCUSSION THREAD

**Discussion by owensgit** [Legal-and-General/canopy#303] [Ideas]
In our util directives such as `lgMargin`, `lgPadding`, `lgCol`, `lgVaraint` etc, we have repeated code that adds/removes classnames to the host element. E.g.

```
  toggleClass(newClass: string, oldClass: string): string {
    if (oldClass) {
      this.renderer.removeClass(this.hostElement.nativeElement, oldClass);
    }
    this.renderer.addClass(this.hostElement.nativeElement, newClass);
    return newClass;
  }
```

Perhaps we should make that more generic and reusable?

Do you agree? If so, please let me know how you think it could be done.

I've added a couple of ideas.

**Comment by owensgit** [comment]
**Base class**
We could create a base class that encapsulates this functionality, requiring derived classes to specify the name of the HTML class name to be added/removed. Or perhaps just inheriting that particular function etc and using it.

**Reply by andy-polhill** [reply]
On the whole I would `favour composition over inheritance`, something like a [mixin based](https://blog.bitsrc.io/composing-angular-components-with-typescript-mixins-31e3833560f7) approach. It would be interesting to see if we can do this and still keep things relatively simple.

**Comment by owensgit** [comment]
**Util method**
One idea is creating a common util method, perhaps on the LgDomService or something like that.
I'm not a fan of this approach as it would require passing in the Renderer and host element into the service so that the classes could be added/removed - seems like the amount of repetition would still be quite high. Would be happy to discuss though, incase I'm missing something.

**Reply by andy-polhill** [reply]
I think we tried something along the lines of a util class, but the code to call the util almost ended up being the same size. I think something maybe changed more recently in Angular that might improve this, let me have a look.

**Comment by owensgit** [comment]
**Don't make more generic**
We could also just keep it as-is and accept that there is some repetition - which might be fine too. Perhaps hold off until we feel there is a stronger case for it.

**Reply by andy-polhill** [reply]
Tricky one, I think it might be worth a play, the code does look repetitive but I think previous attempts to fix this only really added a but more complexity. It's certainly worth having a play round with either approach to see what comes out.

**Comment by andy-polhill** [comment]
Ah I remember now, I think it's linked to [this issue](https://github.com/angular/angular/pull/34938), pre-ivy you couldn't add classes to the host without destroying ones set from else where. Meaning we always had trouble with the way our utility directives played with the component. The component classes took priority and removed the ones set by the directives. I think Ivy altered the order of priority, meaning you can now do the following, Much less code.

```ts
  @Input() showIcon = true;
  @Input() variant: AlertVariant = 'generic';

  @HostBinding('class') classes: string = `lg-alert lg-alert--${this.variant}`;
```

**Reply by andy-polhill** [reply]
(that almost seems too good to be true, definitely worth some checking!)
